### PR TITLE
Update Ballistic animation & ClampingScrollSimulation

### LINKED ISSUE #11599: Ballistic simulations may be created too frequently
Make a normal fling gesture. 
~50% of the time, many goBallistic calls are triggered when I would have expected just one. 

In this example, it took 78ms cumulative recreating the simulations and causes frame drops. 
cw5hukq7xsp

Call stack
```
#0      ScrollPositionWithSingleContext.goBallistic (package:flutter/src/widgets/scroll_position_with_single_context.dart:150:16)
#1      BallisticScrollActivity.applyNewDimensions (package:flutter/src/widgets/scroll_activity.dart:399:14)
#2      ScrollPosition.applyNewDimensions (package:flutter/src/widgets/scroll_position.dart:441:14)
#3      ScrollPositionWithSingleContext.applyNewDimensions (package:flutter/src/widgets/scroll_position_with_single_context.dart:110:29)
#4      ScrollPosition.applyContentDimensions (package:flutter/src/widgets/scroll_position.dart:411:7)
#5      RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:827:20)
#6      RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1767:7)
#7      PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1056:18)
#8      BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:251:19)
#9      BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:504:22)
#10     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:189:5)
#11     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:695:15)
#12     BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:643:9)
#13     _invoke (file:///Users/xster/development/flutter/bin/cache/pkg/sky_engine/lib/ui/hooks.dart:86)
#14     _drawFrame (file:///Users/xster/development/flutter/bin/cache/pkg/sky_engine/lib/ui/hooks.dart:75)
```

### LINKED ISSUE #16371: Android scroll speed deceleration rate doesn't match native
## Steps to Reproduce

1, Scroll a ListView page on Android.  
2, Observe and compare the scrolls between Flutter and Native pages.
3, The scroll is tiny bumpy in Flutter. Besides, before the end of scroll, I have a feeling that it "suddenly" stops.

[screenrecord_20180409_213143.mp4.zip](https://github.com/flutter/flutter/files/1890829/_20180409_213143.mp4.zip)

## Flutter Doctor

[✓] Flutter (Channel beta, v0.2.3, on Mac OS X 10.12.6 16G1314, locale zh-Hans-HK)
    • Flutter version 0.2.3 at /Users/guoyou/projects/flutter
    • Framework revision 5a58b36e36 (4 weeks ago), 2018-03-13 13:20:13 -0700
    • Engine revision e61bb9ac3a
    • Dart version 2.0.0-dev.35.flutter-290c576264

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/guoyou/Library/Android/sdk
    • Android NDK at /Users/guoyou/Library/Android/sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.4.0

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] IntelliJ IDEA Community Edition (version 2017.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 172.3968.27

[✓] Connected devices (1 available)
    • iPhone X • 8D4A21F9-E417-43CE-8B40-52CEAB6B7D04 • ios • iOS 11.2 (simulator)

• No issues found!

### PR DESCRIPTION
**Pull request by Piinks**
Status: _WIP, following up on past review comments and concerns_

Customer impact: ✅  non-breaking

> **Note**
> Part of this change in the past caused a bad regression. This does not appear to be the case anymore with the ballistic animation changes. See here for context: https://github.com/flutter/flutter/pull/77497/files#r705784804

Related PRs:
- #97166
  - #95423
  - #97150
- #77497
- #96512
- #100133

Related Issues:
- Fixes #11599
  - #24715
- Fixes #16371 
- #83632
- #24715
---

### Before 

https://user-images.githubusercontent.com/16964204/179286908-c944de30-4f7b-4411-b1ab-f1d7e9c98e27.mov

---

### After

https://user-images.githubusercontent.com/16964204/179286867-7a9958dc-bd8e-401b-81b7-d6e3cddc70a3.mov

---

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (Piinks) [inline packages/flutter/lib/src/animation/animation_controller.dart:762]: From original PR: This needs to be clearer about behavior when the simulations (new and former) are very different. Is there a jump? Should they merge more gracefully?

Source: https://github.com/flutter/flutter/pull/100133/files#r844470742

Comment (Piinks) [reply packages/flutter/lib/src/animation/animation_controller.dart:762]: Also from original PR: Add more context for how and when this is used.

Source: https://github.com/flutter/flutter/pull/100133/files#r844471138

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (Piinks) [inline packages/flutter/lib/src/animation/animation_controller.dart:764]:
```suggestion
```

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:692]: From original PR: Why not implement now?

I think this is ok to return to. The NestedScrollView handles ballistic scroll activity very differently. It is also a temperamental widget, so I feel that I agree - we should evaluate and update it separately.

Source: https://github.com/flutter/flutter/pull/100133/files#r844471868

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:1373]: Same as above.

File: packages/flutter/lib/src/widgets/scroll_activity.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:56]: From original PR: Add docs for when does it return null and why/what does it mean?

Source: https://github.com/flutter/flutter/pull/100133/files#r844474467

File: packages/flutter/lib/src/widgets/scroll_activity.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:537]: From original PR: These need to be documented

Source: https://github.com/flutter/flutter/pull/100133/files#r844473890

File: packages/flutter/lib/src/widgets/scroll_activity.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:565]: From original PR: Similar concern as above, what does this look like when the animation speeds are vastly different?

Source: https://github.com/flutter/flutter/pull/100133/files#r844476769

**Review by Piinks — COMMENTED** [review]
Checked against internal test suite, this will be able to land without breaking customers. ✅

**Review by Piinks — COMMENTED** [review]
CC-ing interested parties @grouma @mxsoul @goderbauer @nt4f04uNd 

I am still following up on a few things here, but you're feedback is welcome. I have merged the ballistic activity update with the clamping simulation update.

They kind of validate each other. See https://github.com/flutter/flutter/pull/77497/files#r705784804 - this test change (the original regression) is no longer reproducing with them combined.

**Review by grouma — APPROVED** [review]
Fantastic!

It looks like the scrolling is ever so slightly off in your video (but much much better). The scrolling logic should essentially be the same with an equal number of spline samples. I wonder what is causing the minor differences?

Nonetheless, I will be extremely happy when this lands. I'm particularly pleased with the removal of the "snapping" and the addition of [this](https://github.com/flutter/flutter/blob/b1b057e853e89fd36e61e6545f97cb19ed91f00a/packages/flutter/test/widgets/scroll_simulation_test.dart#L27) test.

**Comment by MxSoul** [comment]
Thanks @Piinks 
Due to changes in my own work, I didn't have time to follow up on this PR before. I'm very sorry about that, so it's nice to see you're still following up on this PR.
I would love to see better scrolling on Flutter. I have some free time now, I might be able to help if you need it.

**Review by nt4f04uNd — APPROVED** [review]

File: packages/flutter/test/widgets/scroll_activity_test.dart
Comment (Piinks) [reply packages/flutter/test/widgets/scroll_activity_test.dart:201]: Totally, I still have some copy pasta to clean up. 🙃

**Comment by nt4f04uNd** [comment]
Should this PR also remove [this test](https://github.com/nt4f04uNd/flutter/blob/d237e3ab54853de601c5500ec85ebc2cb017c8aa/packages/flutter/test/widgets/scroll_physics_test.dart#L97)?

See comments here as to why https://github.com/flutter/flutter/pull/97166/files#diff-0181ef59af5a74e790689c74db9264d831b7fcb957688b0424de0da1bf279ca5

Original PR where this was added https://github.com/flutter/flutter/pull/59623/files

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (Piinks) [reply packages/flutter/lib/src/animation/animation_controller.dart:764]: Interesting, I actually just removed an assertion that it is currently animating based on the original PR, see this comment: https://github.com/flutter/flutter/pull/100133/files#r844471344. @MxSoul you mentioned there that the assertion is not necessary, do you remember why?

**Comment by Piinks** [comment]
> Should this PR also remove [this test](https://github.com/nt4f04uNd/flutter/blob/d237e3ab54853de601c5500ec85ebc2cb017c8aa/packages/flutter/test/widgets/scroll_physics_test.dart#L97)?

I don't think so, but I will double check. I remember that test and how it was blocking the page scroll physics update you did @nt4f04uNd. I referenced that above. My plan here is to resolve this, so I can unblock and return to that change. I figure we can revisit that test then, WDYT?

**Comment by Piinks** [comment]
I'll be returning to this next week with updates, I am handling a few fires elsewhere this week. 🔥 😅

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (Piinks) [reply packages/flutter/lib/src/animation/animation_controller.dart:765]: I am adding this assertion back. If we find it is an issue, we can investigate further and remove it if necessary.

File: packages/flutter/lib/src/widgets/scroll_activity.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scroll_activity.dart:59]: Added.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart:692]: Will follow up in https://github.com/flutter/flutter/issues/100748

**Comment by Piinks** [comment]
I am running this through internal testing once more to to ensure these updates won't be reverted. Otherwise I think this is ready now. :)

**Comment by Piinks** [comment]
Bah and merge conflicts just landed, will update again tomorrow

**Comment by Piinks** [comment]
Blocked on https://github.com/flutter/flutter/issues/109797

**Comment by xster** [comment]
😮 super amazing! Thank you Kate!

**Pull request merged**
